src/Microsoft.ML.Core/Data/VBufferMutationContext.cs

```diff
@@ -0,0 +1,148 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    /// <summary>
+    /// Various methods for creating <see cref="VBufferMutationContext{T}"/> instances.
+    /// </summary>
+    public static class VBufferMutationContext
+    {
+        /// <summary>
+        /// Creates a mutation context with the same shape (length and density)
+        /// as the <paramref name="destination"/>.
+        /// </summary>
+        public static VBufferMutationContext<T> CreateFromBuffer<T>(
+            ref VBuffer<T> destination)
+        {
+            return destination.GetMutableContext();
+        }
+
+        /// <summary>
+        /// Creates a mutation context using <paramref name="destination"/>'s values
+        /// and indices buffers.
+        /// </summary>
+        /// <param name="destination">
+        /// The destination buffer.
+        /// </param>
+        /// <param name="newLogicalLength">
+        /// The new length of the buffer being mutated.
+        /// </param>
+        /// <param name="valuesCount">
+        /// The optional number of physical values to be represented in the buffer.
+        /// The buffer will be dense if <paramref name="valuesCount"/> is omitted.
+        /// </param>
+        /// <param name="keepOldOnResize">
+        /// True means that the old buffer values and indices are preserved, if possible (Array.Resize is called).
+        /// False means that a new array will be allocated, if necessary.
+        /// </param>
+        /// <param name="requireIndicesOnDense">
+        /// True means to ensure the Indices buffer is available, even if the buffer will be dense.
+        /// </param>
+        public static VBufferMutationContext<T> Create<T>(
+            ref VBuffer<T> destination,
+            int newLogicalLength,
+            int? valuesCount = null,
+            bool keepOldOnResize = false,
+            bool requireIndicesOnDense = false)
+        {
+            return destination.GetMutableContext(
+                newLogicalLength,
+                valuesCount,
+                keepOldOnResize: keepOldOnResize,
+                requireIndicesOnDense: requireIndicesOnDense);
+        }
+
+        internal static VBufferMutationContext<T> Create<T>(
+            ref VBuffer<T> destination,
+            int newLogicalLength,
+            int valuesCount,
+            int maxValuesCapacity)
+        {
+            return destination.GetMutableContext(
+                newLogicalLength,
+                valuesCount,
+                maxValuesCapacity);
+        }
+    }
+
+    /// <summary>
+    /// An object capable of mutation a <see cref="VBuffer{T}"/> by filling out
+    /// <see cref="Values"/> (and <see cref="Indices"/> if the buffer is not dense).
+    /// </summary>
+    public ref struct VBufferMutationContext<T>
```

Comment:
>VBufferMutationContext [](start = 22, length = 22)

I'm not a big fan of this name. Is the discussion closed on naming? 
Given the usage pattern, maybe we can call it `VBufferBuilder` or `VBufferEditor`?

Reply:
well, well. I didn't actually understand what you're planning to do until I saw it implemented.

`BufferBuilder` is internal anyway, right?

---
In reply to: [233672374](https://github.com/dotnet/machinelearning/pull/1580#discussion_r233672374) [](ancestors = 233672374)

Reply:
I would much rather rename `BufferBuilder` to something ridiculous like 'value accumulator', and have the good name for the more common (and public) use case.

---
In reply to: [233675246](https://github.com/dotnet/machinelearning/pull/1580#discussion_r233675246) [](ancestors = 233675246,233672374)

Reply:
> BufferBuilder is internal anyway, right?

It isn't right now.

https://github.com/dotnet/machinelearning/blob/2d2ed27747ae7bd89723f53811fa6b12f1b28816/src/Microsoft.ML.Data/Data/BufferBuilder.cs#L17

I don't really view this thing as a "builder" anyway. For one there is no "Build" method on it.  `VBufferEditor` would be more appropriate IMO. Do you like that?

* VBufferMutation
* VBufferMutationContext
* VBufferEditor

Any others to consider?